earth_enterprise/src/fusion/autoingest/sysman/StateUpdater.cpp

```diff
@@ -278,16 +291,76 @@ class StateUpdater::SetStateVisitor : public default_dfs_visitor {
     }
 };
 
-void StateUpdater::SetStateForRefAndDependents(
+
+class StateUpdater::SetBlockingStateVisitor : public StateUpdater::VisitorBase {
+  private:
+    using BlockedSet = std::unordered_set<SharedString>;
+
+    // This is a shared_ptr because the visitor will be copied several times
+    // during the course of the traversal.
+    const std::shared_ptr<BlockedSet> hasBlockingInputs;
+    const std::shared_ptr<BlockedSet> hasBlockingChildren;
+
+    void MarkParentsForLaterProcessing(AssetHandle<AssetVersionImpl> & version) const override {
+      if (AssetDefs::IsBlocking(newState)){
+        hasBlockingChildren->insert(version->parents.begin(), version->parents.end());
+        hasBlockingInputs->insert(version->listeners.begin(), version->listeners.end());
+      }
+    }
+
+  public:
+    SetBlockingStateVisitor(StateUpdater & updater, DependentStateTree & tree, AssetDefs::State newState) :
+        VisitorBase(updater, tree, newState),
+        hasBlockingInputs(std::make_shared<BlockedSet>()),
+        hasBlockingChildren(std::make_shared<BlockedSet>()) {}
+
+    // This function is called after the DFS has completed for every vertex
+    // below this one in the tree. Thus, we don't calculate the state for an
+    // asset until we've calculated the state for all of its inputs and
+    // children.
+    void finish_vertex(
+        DependentStateTreeVertexDescriptor vertex,
+        const DependentStateTree &) const {
+      const AssetVertex & data = tree[vertex];
+      notify(NFY_PROGRESS, "Calculating state for '%s'", data.name.toString().c_str());
+
+      // Set the state for assets in the dependent tree.
+      if (data.inDepTree) {
+        SetState(vertex, newState, {0, 0}, true);
+      }
+      else {
+        // Set the state to blocked if needed, otherwise do nothing
+        auto version = updater.storageManager->Get(data.name);
+        if (hasBlockingInputs->find(data.name) != hasBlockingInputs->end() && 
+            version->InputStatesAffectMyState(AssetDefs::Blocked, true)) {
+          SetState(vertex, AssetDefs::Blocked, {0,0}, true);
+        }
+        else if (hasBlockingChildren->find(data.name) != hasBlockingChildren->end()) {
+          SetState(vertex, AssetDefs::Blocked, {0,0}, true);
+        }
+      }
+    }
+};
+
+/**********************************************
+* StateUpdater
+**********************************************/
+void StateUpdater::SetAndPropagateState(
     const SharedString & ref,
     AssetDefs::State newState,
     function<bool(AssetDefs::State)> updateStatePredicate) {
   try {
     SharedString verref = AssetVersionImpl::Key(ref);
-    bool includeDepDescendents = !UserActionRequired(newState);
+    bool includeAllChildrenAndInputs = !UserActionRequired(newState);
```

Comment:
There's a small bug here: if `newState` is `Failed`, we want `includeAllChildrenAndInputs` to be false. Basically, any time we use a blocking state visitor we want includeAllChildrenAndInputs to be false. The way it is now isn't wrong, but it does unnecessary work.

Reply:
Good catch. I'll update that.

Reply:
Both `includeAllChildrenAndInputs` and `includeDependentChildren` were going to have to check for `newState` being `Failed`, so I just decided to make it a separate if statement. Not sure which way is cleaner.